File: c3o.Logger.Web/app/core/models/LogMessage.js
/////////////////////////////////////////////////////////////
// LogMessage Class
/////////////////////////////////////////////////////////////
c3o.Core.Data.LogMessage = function (data, model) {
	var self = this;

	_.extend(this, data);
		
	this.log = this.log || _.find(model.logs, { id: this.logId });
	this.application = this.application || _.find(model.applications, { id: this.applicationId });
	this.user = this.user || _.find(model.users, { id: this.userId });
	this.messageType = this.messageType || _.find(model.types, { id: this.messageTypeId });
	this.source = this.source || _.find(model.sources, { id: this.sourceId });
	var severityObj = _.find(model.severities, { name: this.severity });

	if (this.log) {
		this.log.messages = this.log.messages || [];
		this.log.messages.push(this);
	};

	if (this.application) {
		this.application.messages = this.application.messages || [];
		this.application.messages.push(this);
	};

	if (this.user) {
		this.user.messages = this.user.messages || [];
		this.user.messages.push(this);
	};

	if (this.messageType) {
		this.messageType.messages = this.messageType.messages || [];
		this.messageType.messages.push(this);
	};

	if (this.source) {
		this.source.messages = this.source.messages || [];
		this.source.messages.push(this);
	};

	if (severityObj) {
		severityObj.messages = severityObj.messages || [];		
		severityObj.messages.push(this);
	};
}

// SiteContent class methods
c3o.Core.Data.LogMessage.prototype = {
	//get Name() { return this.firstName + " " + this.lastName; },
	//get NameLastFirst() { return this.lastName + ", " + this.firstName; }
	
}

File: Elmah.Logger.Data/Model/LogMessage.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using Elmah.Io.Client;
using Newtonsoft.Json;
using c3o.Core;

namespace c3o.Logger.Data 
{
	//public enum LogSeverity { Verbose = 0, Debug = 1, Information = 2, Warning = 3, Error = 4, Fatal = 5, }

    public class LogMessage 
	{
		public long Id { get; set; }
		public string ElmahId { get; set; }
		public DateTime DateTime { get; set; }
        public string Hostname { get; set; }		
		public LogSeverity Severity { get; set; }
        public int? StatusCode { get; set; }
		public string Title { get; set; }
		public string Url { get; set; }
		public string Blob { get; set; }
		public long? IpAddressId { get; set; }
        public long LogId { get; set; }
        public long? ApplicationId { get; set; }
        public long? UserId { get; set; }
        public long? MessageTypeId { get; set; }
        public long? SourceId { get; set; }
        public long? DetailId { get; set; }
        public int LogCount { get; set; }

        public Log Log { get; set; }
		public LogApplication Application { get; set; }
		public LogUser User { get; set; }
		public LogMessageType MessageType { get; set; }
		public LogMessageSource Source { get; set; }		
		public LogMessageDetail Detail { get; set; }

        [NotMapped]
        public string IpAddress
        {
            get { return this.IpAddressId == null ? null : NetworkHelper.IPv4Int64ToString(this.IpAddressId.Value); }
            set { this.IpAddressId = NetworkHelper.IPv4StringToInt64(value); }
        }

        [NotMapped]
		public string HowLongAgo { get { return this.DateTime.GetDelta(); } }

		[NotMapped]
		public LogMessageBlob Original { get { return JsonConvert.DeserializeObject<LogMessageBlob>(this.Blob); } }

		public LogMessage()
        {
            this.LogCount = 1;
        }
		public LogMessage(Message obj, string logId) : this()
        {
			this.ElmahId = obj.Id;
			if (string.IsNullOrWhiteSpace(this.ElmahId)) { this.ElmahId = obj.Data.GetValue("ElmahId"); }
			
			this.DateTime = DateTime.UtcNow; // Use UTC server dt
			this.Hostname = obj.Hostname;
            this.Severity = (LogSeverity)(obj.Severity ?? Elmah.Io.Client.Severity.Error);
            this.StatusCode = obj.StatusCode;
			this.Title = obj.Title;
			
			this.Url = obj.Url;
		    //if (string.IsNullOrWhiteSpace(this.Url)) { this.Url = obj.Data.GetValue("Url"); }
			//if (string.IsNullOrWhiteSpace(this.Url)) { this.Url = obj.ServerVariables.GetValue("Url"); }
			
			var ip = obj.ServerVariables.GetValue("REMOTE_ADDR");
			if (!string.IsNullOrWhiteSpace(ip))
			{
				this.IpAddress = ip;
			}

			//this.Type = obj.Type;
			this.Blob = JsonConvert.SerializeObject(new LogMessageBlob(obj));

			// log & application
			if (!string.IsNullOrWhiteSpace(obj.Application))
			{
				if (this.Log == null)
				{
					this.Log = new Log() { LogId = new Guid(logId), Name = obj.Application };
				}
				if (this.Application == null)
				{
					this.Application = new LogApplication() { Name = obj.Application };
				}				
			}

			// source
			if (!string.IsNullOrWhiteSpace(obj.Source))
			{
				if (this.Source == null)
				{
					this.Source = new LogMessageSource { Name = obj.Source };
				}
			}

			// type
			if (!string.IsNullOrWhiteSpace(obj.Type))
			{
				if (this.MessageType == null)
				{
					this.MessageType = new LogMessageType { Name = obj.Type };
				}
			}			

			// user
			if (!string.IsNullOrWhiteSpace(obj.User))
			{
				if (this.User == null)
				{
					this.User = new LogUser() { Name = obj.User };
				}
			}
			
			// detail
			if (!string.IsNullOrWhiteSpace(obj.Detail))
			{
				if (this.Detail == null)
				{
					this.Detail = new LogMessageDetail() { 
						Content = obj.Detail,
						ContentKey = StringHelper.CheckSum(obj.Detail)
					};
				}
			}			
		}
	}
}
